prototype_source/vulkan_workflow.rst

```diff
@@ -0,0 +1,187 @@
+PyTorch Vulkan Backend User Workflow
+====================================
+
+**Author**: `Ivan Kobzarev <https://github.com/IvanKobzarev>`_
+
+Introduction
+------------
+PyTorch 1.7 supports the ability to run model inference on GPUs that support the Vulkan graphics and compute API. The primary target devices are mobile GPUs on Android devices. The Vulkan backend can also be used on Linux, Mac, and Windows desktop builds to use Vulkan devices like Intel integrated GPUs. This feature is in the prototype stage and is subject to change.
+
+Building PyTorch with Vulkan backend
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Vulkan backend is not included by default. The main switch to include Vulkan backend is cmake option ``USE_VULKAN``, that can be set by environment variable ``USE_VULKAN``.
+
+To use PyTorch with Vulkan backend, we need to build it from source with additional settings. Checkout the PyTorch source code from GitHub master branch.
+
+Vulkan wrapper
+^^^^^^^^^^^^^
+
+By default, Vulkan library will be loaded at runtime using the vulkan_wrapper library. If you specify the environment variable ``USE_VULKAN_WRAPPER=0`` libvulkan will be linked directly.
+
+Desktop build
+^^^^^^^^^^^^^
+
+Download VulkanSDK from https://vulkan.lunarg.com/sdk/home and set environment variable ``VULKAN_SDK``
+
+For Linux:
+
+::
+
+    cd PYTORCH_ROOT
+    USE_VULKAN=1 USE_VULKAN_SHADERC_RUNTIME=1 USE_VULKAN_WRAPPER=0 python setup.py install
+
+For Mac:
+
+::
+
+    cd PYTORCH_ROOT
+    USE_VULKAN=1 USE_VULKAN_SHADERC_RUNTIME=1 USE_VULKAN_WRAPPER=0 MACOSX_DEPLOYMENT_TARGET=10.9 CC=clang CXX=clang++ python setup.py install
+
+After successful build, open another terminal and verify the version of installed PyTorch.
+
+::
+
+    import torch
+    print(torch.__version__)
+
+At the time of writing of this recipe, the version is 1.8.0a0+41237a4. You might be seeing different numbers depending on when you check out the code from master, but it should be greater than 1.7.0.
+
+
+Android build
+^^^^^^^^^^^^^
+::
+
+    cd PYTORCH_ROOT
+    ANDROID_ABI=arm64-v8a USE_VULKAN=1 sh ./scripts/build_android.sh
+
+
+Model preparation
+-----------------
+
+Install torchvision, get the default pretrained float model.
+
+::
+
+    pip install torchvision
+
+Python script to save pretrained mobilenet_v2 to a file:
+
+::
+
+    import torch
+    import torchvision
+
+    model = torchvision.models.mobilenet_v2(pretrained=True)
+    model.eval()
+    script_model = torch.jit.script(model)
+    torch.jit.save(script_model, “mobilenet2.pt”)
+
+PyTorch 1.7 Vulkan backend supports only float 32bit operators. The default model needs additional step that will optimize operators fusing 
+
+::
+
+    from torch.utils.mobile_optimizer import optimize_for_mobile
+    script_model_vulkan = optimize_for_mobile(script_model, backend=’Vulkan’)
+    torch.jit.save(script_model_vulkan, "mobilenet2-vulkan.pt")
+
```

Comment:
I ran the script above and got a long error message (in attachment). The complete script I used is:
```
import torch
import torchvision

model = torchvision.models.mobilenet_v2(pretrained=True)
model.eval()
script_model = torch.jit.script(model)
torch.jit.save(script_model, "mobilenet2.pt")

from torch.utils.mobile_optimizer import optimize_for_mobile
script_model_vulkan = optimize_for_mobile(script_model, backend='Vulkan')
torch.jit.save(script_model_vulkan, "mobilenet2-vulkan.pt")
```